qlight/token_holder.go

```diff
@@ -1,11 +1,177 @@
+	plugin, err := getPlugin(pluginManager)
+	if err != nil {
+		return nil, fmt.Errorf("get plugin: %w", err)
+	}
+	var refreshAnticipation int32
+	if plugin != nil {
+		refreshAnticipation, err = plugin.PluginTokenManager(context.Background())
+		if err != nil {
+			return nil, fmt.Errorf("fetch refresh anticipation value: %w", err)
+		}
+	}
+	return NewTokenHolderWithPlugin(psi, refreshAnticipation, peerUpdater, plugin, pluginManager), nil
+}
+
+func getPlugin(pluginManager *plugin.PluginManager) (plugin qlightplugin.PluginTokenManager, err error) {
+	pluginTemplate, err := tokenManager(pluginManager)
+	if err != nil {
+		return
+	}
+	if pluginTemplate != nil {
+		err = pluginTemplate.Start()
+		if err != nil {
+			return
+		}
+		plugin, err = pluginTemplate.Get()
+	}
+	return
+}
+
+func NewTokenHolderWithPlugin(psi string, refreshAnticipation int32, peerUpdater RunningPeerAuthUpdater, plugin qlightplugin.PluginTokenManager, pluginManager *plugin.PluginManager) *TokenHolder {
+	return &TokenHolder{
+		psi:                 psi,
+		plugin:              plugin,
+		pluginManager:       pluginManager,
+		peerUpdater:         peerUpdater,
+		refreshAnticipation: refreshAnticipation,
+	}
+}
+
+func (h *TokenHolder) RefreshPlugin(pluginManager *plugin.PluginManager) (err error) {
+	h.plugin, err = getPlugin(pluginManager)
+	if err != nil {
+		return
+	}
+	var refreshAnticipation int32
+	if h.plugin != nil {
+		refreshAnticipation, err = h.plugin.PluginTokenManager(context.Background())
+		if err != nil {
+			return
+		}
+	}
+	h.refreshAnticipation = refreshAnticipation
+	return
+}
+
+func (h *TokenHolder) HttpCredentialsProvider(ctx context.Context) (string, error) {
+	if h.plugin != nil {
+		log.Debug("HttpCredentialsProvider using plugin")
+		return h.plugin.TokenRefresh(ctx, h.token, h.psi)
+	}
+	log.Debug("HttpCredentialsProvider using token")
+	return h.token, nil
+}
+
+func (h *TokenHolder) ReloadPlugin() error {
+	plugin, err := getPlugin(h.pluginManager)
+	if err != nil {
+		return err
+	}
+	refreshAnticipation, err := plugin.PluginTokenManager(context.Background())
+	if err != nil {
+		return fmt.Errorf("fetch refresh anticipation value: %w", err)
+	}
+	h.plugin = plugin
+	h.refreshAnticipation = refreshAnticipation
+	return nil
+}
+
+func (h *TokenHolder) CurrentToken() string {
+	if h == nil {
+		log.Warn("token holder nil, returning empty token")
+		return ""
+	}
+	if h.plugin == nil {
+		log.Warn("token plugin is missing, no update possible")
+		return h.token
+	}
+	expired, err := h.tokenExpired()
+	if err != nil {
+		log.Warn("error while checking if token is expired", "err", err)
+	}
+	if !expired {
+		return h.token
+	}
+	returnedToken, err := h.plugin.TokenRefresh(context.Background(), h.token, h.psi)
+	if err != nil {
+		log.Error("get token from plugin", "err", err)
+	} else {
+		if h.token != returnedToken {
+			log.Debug("new token from plugin")
+			err = h.peerUpdater.UpdateTokenForRunningQPeers(returnedToken)
```

Comment:
From what I understand if you keep calling `CurrentToken()` (most likely due to RPC requests that must be forwarded to the qlight server node) the corresponding token will also be updated in the P2P connection (whenever it is refreshed).
What happens if the qlight client node DOES NOT make any RPC requests (sendTransaction*) that are forwarded to the qlight server? 
I suspect the qlight server will close the P2P connection because the token expires and when a new connection is established the qlight client will get a new token. Is that the expected behaviour?
I was kind of expecting some timer logic to prevent the P2P connection from being dropped.

Reply:
I will add a timer that will do a default refresh. Thank you for raising this.

Reply:
I've updated the code

Reply:
LGTM